### DISCUSSION TITLE
Import .wasm files

### DISCUSSION THREAD

**Discussion by tel** [quarto-dev/quarto-cli#6494] [Feature Requests]
### Description

It'd be excellent to be able to import .wasm modules in ojs blocks. Currently there's an "Unknown language in file ..." error that arises when importing a wasm module or even a Javascript file providing a shim around a wasm module.

```javascript
import { greet } from "./pkg/wasm_tut.js"
```

```javascript
// wasm_tut.js
import * as wasm from "./wasm_tut_bg.wasm"; // likely what is causing the error
import { __wbg_set_wasm } from "./wasm_tut_bg.js";
__wbg_set_wasm(wasm);
export * from "./wasm_tut_bg.js";
```

My particular use case is to be able to import models that I've written in Rust into Quarto in order to execute them interactively. To that end, I'm using the `wasm-pack` tool to output `.wasm` files and auto-generated shim `.js` modules.

Thanks!

**Comment by mcanouil** [comment]
This is a question for Observable.
Quarto team has no control on what Observable can or cannot do.

**Reply by tel** [reply]
I agree the import mechanism may not directly load local packages and that’s on Observable. What comes to mind is that it’d be nice to be able to configure the location of a package by name, similar to webpack’s “packages” config.

I could then `require(“mypkg”)` and it would not attempt to download it from jsdeliver but instead use the local package.

I believe the resolver used by the require function is injected by the user of Observable’s Library. So I think this could be done via Quarto but I’m not entirely confident.

I’m also going to experiment with putting the wasm module directly into node_modules though this is a bit undesirable over having this explicitly in configuration.

**Reply by mcanouil** [reply]
@cscheid is the one that could help here but he is not available for a few days.

**Reply by tel** [reply]
Thanks for making the connection!

**Comment by tel** [comment]
A bit more thinking about this, the modules are loaded browserside so there's no `node_modules` directory. I tried using another webserver to host the module, but this is impeded by a CORS policy. I could imagine this being possible through Quarto using the `FileAttachment` technology.

**Reply by tel** [reply]
My first attempt

```js
pkg = import(await FileAttachment("pkg/wasm_tut.js").url())
```

fails with error `wasm_tut_bg.wasm:1 Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "application/wasm". Strict MIME type checking is enforced for module scripts per HTML spec.`

**Reply by tel** [reply]
Ah, this works, I just had to properly configure `wasm-pack` to export the module as a web module. It's not a particularly obvious method, I suppose, but it might be nice to document this.

**Reply by mcanouil** [reply]
In a Quarto extension you could possibly add those resources as it is done by https://github.com/coatless/quarto-webr

**Reply by tel** [reply]
Oh that's an interesting method, too. I'll investigate that.

**Reply by cscheid** [reply]
@tel This is fairly deep waters you're wading into. I believe the `import` syntax you're using is non-standard, and needs to be translated by something like babel to "proper" javascript. See https://developer.mozilla.org/en-US/docs/WebAssembly/Loading_and_running. Currently, the {OJS, TS} -> JS translation we do uses the tooling available in Deno, `acornjs` and `esbuild`. The maintenance burden on this is fairly high, so we have no plans to introduce additional heavyweight machinery like webpack, babel, etc. If you want to use those, I recommend you create a pre-render script that compiles those down to "plain" ESM modules, and import that instead.

**Reply by tel** [reply]
Definitely looking for whatever is easiest. I reached for the async `import` technology because I saw reference to it from mbostock somewhere and I got it working. The tool I'm using to generate the wasm file is `wasm-pack` and it offers different sorts of packaging: https://rustwasm.github.io/docs/wasm-bindgen/reference/deployment.html

For instance, I can ask it to export a `nodejs` style module (using `module.exports`) but I cannot get the `require` function to accept it

```js
pkg = require(await FileAttachment("pkg/model.js").url());
```

fails with `RequireError: invalid module`. It mostly seems upset because `module` doesn't exist in a line like `module.exports = {}`.

**Reply by cscheid** [reply]
I think OJS's `require` expects an ESM module instead. have you tried setting everything up with in a small, handwritten .js file first to make sure you know what's happening at every step of the way? `wasm-pack` will add a lot of extra layers of complexity, and I think you'd benefit from making sure you understand what's happening in simpler cases first.

**Reply by tel** [reply]
I hadn't yet, I was trying to get my head around how the various larger scale pieces were working. I'm also presently satisfied with using the `import` statement, though it's likely not very portable.

**Reply by younes-io** [reply]
@tel I'd like to also import `wasm` files into quarto.. Could you please share how did you manage to do that ? Really appreciate it!
@cscheid : if you came up, since then, with a better way, could you please share ?
Thank you all

**Reply by tel** [reply]
I ended up writing a block like this

```js
pkg = {
  const file = await FileAttachment("pkg/model.js").url();
  const module = await import(file);
  module.default = await module.default();
  module.set_panic_hook();
  return module;
}
```

where `pkg` contains the output from calling warm-pack

```
wasm-pack build -t web -d ../pkg/
```

**Reply by younes-io** [reply]
okay, thank you
